File: 10-Stack-Queue-PriorityQueue/04-Priority-Queue-Heap/cpp/top_k_frequent.cpp
/*
 * Problem: Top K Frequent Elements
 * Source:  LeetCode 347
 * Time: O(n log k), Space: O(n)
 * Tags: priority-queue, heap, hashing
 */
#include <bits/stdc++.h>
using namespace std;

vector<int> topKFrequent(vector<int>& nums, int k) {
    unordered_map<int, int> freq;
    for (int x : nums) freq[x]++;
    
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> minHeap;
    for (auto& [val, cnt] : freq) {
        minHeap.push({cnt, val});
        if (minHeap.size() > k) minHeap.pop();
    }
    
    vector<int> result;
    while (!minHeap.empty()) { result.push_back(minHeap.top().second); minHeap.pop(); }
    return result;
}

int main() {
    vector<int> nums = {1,1,1,2,2,3};
    auto res = topKFrequent(nums, 2);
    for (int x : res) cout << x << " "; // 2 1 (or 1 2)
    return 0;
}

File: 10-Stack-Queue-PriorityQueue/04-Priority-Queue-Heap/java/TopKFrequent.java
import java.util.*;
public class TopKFrequent {
    public static int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int x : nums) freq.merge(x, 1, Integer::sum);
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        for (var e : freq.entrySet()) {
            minHeap.offer(new int[]{e.getKey(), e.getValue()});
            if (minHeap.size() > k) minHeap.poll();
        }
        int[] result = new int[k]; int i = 0;
        while (!minHeap.isEmpty()) result[i++] = minHeap.poll()[0];
        return result;
    }
    public static void main(String[] args) {
        for (int x : topKFrequent(new int[]{1,1,1,2,2,3}, 2)) System.out.print(x + " ");
    }
}
